File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/MusicSequenceNode.cs
using System.Collections.Generic;
using SoundPlayground.Core;
using SoundPlayground.VirtualMachine;

namespace SoundPlayground.Parser.AbstractSyntaxTree
{
    public class MusicSequenceNode : MusicNode
    {
        public IEnumerable<MusicNode> Expressions { get; set; }

        public MusicSequenceNode ( IEnumerable<MusicNode> nodes ) {
            Expressions = nodes;
        }

        public override IEnumerable<Note> GetCommands ( Context context ) {
            foreach ( MusicNode node in Expressions ) {
                foreach ( Note note in node.GetCommands( context ) ) {
                    yield return note;
                }
            }
        }
    }
}

File: code/musikla/musikla/parser/abstract_syntax_tree/music_sequence_node.py
from musikla.parser.printer import CodePrinter
from typing import Tuple
from musikla.core import Value, Music
from .music_node import MusicNode

class MusicSequenceNode( MusicNode ):
    def __init__ ( self, nodes, position : Tuple[int, int] = None ):
        super().__init__( position )

        self.expressions = nodes

    def to_source ( self, printer : CodePrinter ):
        for i in range( len( self.expressions ) ):
            if i > 0: printer.add_token( ' ' )

            self.expressions[ i ].to_source( printer )

    def get_events ( self, context ):
        for node in self.expressions:
            value = node.eval( context )

            if isinstance( value, Music ):
                for event in value.expand( context ):
                    yield event

                    if event.end_timestamp > context.cursor:
                        context.cursor = event.end_timestamp
    
    def __iter__ ( self ):
        return iter( self.expressions )
